### ISSUE TITLE
Lower precision for GRU4Rec compared to NeuMF on ML-100k dataset using full sort evaluation

### ISSUE THREAD

**Issue by skcgd** [RUCAIBox/RecBole#1311]
Hi,
When I compared GRU4Rec with NeuMF on ML-100k dataset using the full sort mode, I found the precison@5 and precions@10 for GRU4Rec was about 0.008 compared to 0.02 of NeuMF. Turning parameters does not change the result much. Are there some bugs? Thanks.

**Comment by hyp1231** [comment]
Hi, it is because that, in RecBole, sequential models (e.g., GRU4Rec) are in "repeatable recommendation" mode by default (can recommend items that have been interacted in training set while evaluation), while general models (e.g., NeuMF) are in "non-repeatable recommendation" mode.

For a fair comparison, please set `repeatable: True` when runing NeuMF. [[API Doc]](https://recbole.io/docs/user_guide/config/evaluation_settings.html?highlight=repeatable)

**Comment by skcgd** [comment]
@hyp1231 I have already set the repeatable: True when running NeuMF. For other parameters, I used the default setting.

**Comment by hyp1231** [comment]
Hi @skcgd,

Besides unifying repeatable recommendation mode, we should also consider the evaluation setting. Because if we split the dataset by ratio (e.g., 8:1:1 by default), the test set for GRU4Rec and NeuMF differs from each other. When we split the datasets by ratio, each user may have multiple ground truth items in valid/test set, while only one ground truth item for each sequence. Then the metrics comparison can make no sense as the test sets of these two models are different.

Thus, we suggest to split the datasets in a "leave one out" mode. In this way, only one ground truth item for each user will exist in valid/test sets, making the comparison fair. By the way, most sequential models are evaluated in leave one out mode in their original papers.

I've also tried it, by using the following config file, both sample one negative sample for each positive item, and split the datasets in leave one out mode, 
```yaml
# dataset config
load_col:
  inter: [user_id, item_id, rating, timestamp]
val_interval:
  rating: "[3,inf)"

# training and evaluation
train_batch_size: 4096
valid_metric: recall@10
eval_args:
  split: {'LS':"valid_and_test"}
  mode: full
  order: TO
neg_sampling: {'uniform': 1}
repeatable: True
loss_type: BPR
```

we have
```
OrderedDict([('recall@10', 0.0456), ('mrr@10', 0.0111), ('ndcg@10', 0.019), ('hit@10', 0.0456), ('precision@10', 0.0046)])
```
for NeuMF, and
```
OrderedDict([('recall@10', 0.053), ('mrr@10', 0.0144), ('ndcg@10', 0.0234), ('hit@10', 0.053), ('precision@10', 0.0053)])
```
for GRU4Rec

**Comment by Sherry-XLL** [comment]
Hello @skcgd!

The suggestions and feedback from open source users are our driving force for continuous updating. We welcome you very much to use our library again and give your valuable comments.

Closing due to inactivity. Please open a new issue if you're still having questions.
